src/Query/ConditionGroup.php

```diff
@@ -0,0 +1,194 @@
+<?php
+
+namespace Drupal\entity\Query;
+
+use Drupal\Core\Cache\Cache;
+use Drupal\Core\Cache\CacheableDependencyInterface;
+use Drupal\Core\Cache\RefinableCacheableDependencyTrait;
+
+/**
+ * Represents a group of access conditions.
+ *
+ * Used by query access handlers for filtering lists of entities based on
+ * granted permissions.
+ *
+ * Examples:
+ * @code
+ *   // Filter by node type and uid.
+ *   $condition_group = new ConditionGroup();
+ *   $condition_group->addCondition('type', ['article', 'page']);
+ *   $condition_group->addCondition('uid', '1');
+ *
+ *   // Filter by node type or status.
+ *   $condition_group = new ConditionGroup('OR');
+ *   $condition_group->addCondition('type', ['article', 'page']);
+ *   $condition_group->addCondition('status', '1', '<>');
+ *
+ *   // Nested condition groups: node type AND (uid OR status).
+ *   $condition_group = new ConditionGroup();
+ *   $condition_group->addCondition('type', ['article', 'page']);
+ *   $condition_group->addCondition((new ConditionGroup('OR'))
+ *     ->addCondition('uid', 1)
+ *     ->addCondition('status', '1')
+ *   );
+ * @endcode
+ */
+final class ConditionGroup implements \Countable, CacheableDependencyInterface {
+
+  use RefinableCacheableDependencyTrait;
+
+  /**
+   * The conditions.
+   *
+   * @var \Drupal\entity\Query\Condition[]|\Drupal\entity\Query\ConditionGroup[]
+   */
+  protected $conditions = [];
+
+  /**
+   * The conjunction.
+   *
+   * @var string
+   */
+  protected $conjunction;
+
+  /**
+   * Constructs a new ConditionGroup object.
+   *
+   * @param string $conjunction
+   *   The conjunction.
+   */
+  public function __construct($conjunction = 'AND') {
+    $this->conjunction = $conjunction;
+  }
+
+  /**
+   * Gets the conjunction.
+   *
+   * @return string
+   *   The conjunction. Possible values: AND, OR.
+   */
+  public function getConjunction() {
+    return $this->conjunction;
+  }
+
+  /**
+   * Adds a condition.
+   *
+   * @param string|\Drupal\entity\Query\ConditionGroup $field
+   *   Either a condition group (for nested AND/OR conditions), or a
+   *   field name with an optional column name. E.g: 'uid', 'address.locality'.
+   * @param mixed $value
+   *   The value.
+   * @param string $operator
+   *   The operator.
+   *   Possible values: =, <>, <, <=, >, >=, BETWEEN, NOT BETWEEN,
+   *                   IN, NOT IN, IS NULL, IS NOT NULL.
+   *
+   * @return $this
+   */
+  public function addCondition($field, $value = NULL, $operator = NULL) {
+    if ($field instanceof ConditionGroup) {
+      if ($field->count() === 1) {
+        // The condition group only has a single condition, merge it.
+        $this->conditions[] = reset($field->getConditions());
+        $this->addCacheTags($field->getCacheTags());
+        $this->addCacheContexts($field->getCacheContexts());
+        $this->mergeCacheMaxAge($field->getCacheMaxAge());
```

Comment:
There's a method for this: `$this->addCacheableDependency($field);`?

Reply:
Actually, I'm pretty sure this just isn't needed at all because it's always going to be computed below. This is just doing extra work.